### ISSUE TITLE
Code for prostate application ?

### ISSUE THREAD

**Issue by mathilde-b** [neerakara/test-time-adaptable-neural-networks-for-domain-generalization#3]
Hello Neerav, 
First of all, thank you for the very nice contribution and for making your code publicly available. I have a question : it seems that all of the code available is very tailored to the brain application. I am super interested in the prostate application, as I will be reproducing it for my research. If you have this at hand, it would be super useful !

Thank you
Best,
Mathilde

**Comment by neerakara** [comment]
Hi Mathilde,

Thanks a lot for your interest in this work and for reaching out.

1. The implementation for the prostate should be the same, apart from two differences:
    a. for prostate, we could train the 3D DAE in full resolution and did not have to sub-sample along the z-direction as is done for the brain experiments. 
    b. for prostate, all the datasets have a 'small' domain shift. That is, they are images from the same modality, but acquired in different hospitals. So, we do not have the atlas based initialization for test-time adaptation (this had to be done for T2w to T1w adaptation of brain scans).

2. I had also added the prostate code in another repo: https://github.com/neerakara/TTA_prostate. 

Please do let me know if you have further questions about the code - I will be glad to clarify.

Best regards,
Neerav

**Comment by mathilde-b** [comment]
Hello again,
thanks for the quick response ! I'm closer now, but still struggling to run the baseline - I've never worked with tensorflow. Could you give the specifications associated, ie the tensorflow version etc..? Is seems the script is either running on CPU (tf <2) or I can't make it work due to tf.contrib...
Thanks
Mathilde

**Comment by neerakara** [comment]
The code has been tested using Tensorflow v1.12 and runs on GPUs.

**Comment by mathilde-b** [comment]
OK, as I have CUDA 11.4 I need to use tensorflow v2. I was able to pass your code to make it work with tensorflow v2 using this link : https://www.tensorflow.org/guide/migrate/upgrade , it seems to be running on GPU now, and to be working. I'll keep you updated ! Let me know if you know of specific pitfalls I should be aware of.
Merci !
Mathilde

**Comment by mathilde-b** [comment]
Hello Neerav, I'm having a problem with train_i2l_mapper.py : 
For a small number of slices, everything works well.
When I use all slices that I have, the logits and the loss are NAN after a few mini-batch iterations... I really don't understand what's going on here. Any idea ? The data was properly created, there is no NA in it.
thanks !

**Comment by mathilde-b** [comment]
<img width="662" alt="Screenshot 2021-12-04 at 18 31 51" src="https://user-images.githubusercontent.com/25257819/144727798-a9615d86-b16c-41fa-b60c-47e717443d8e.png">

**Comment by neerakara** [comment]
Hi Mathilde,

Could it be that a particular mini-batch consists of all blank slices, which could potentially lead to Nan values (perhaps in the batch normalization layers)? From the top of my head, I am unable to think of something else that could be causing this.

If you think it would be helpful, I would be happy to have a zoom call to debug this together.

Best,
Neerav

**Comment by mathilde-b** [comment]
Hey Neerav,
Ok so actually I was wrong, it does seem that quite a lot of NAN values are generated when creating the hdf5 file in the data/my_data.py that I adapted from yours... I don't really understand why, because I checked and the counter_from does not seem to skip any iteration... Do you have any idea why this could happen?
Yes we could do a zoom to debug this !

**Comment by neerakara** [comment]
This seems to be a data related issue. I have never encountered this while working with several datasets. Which dataset are you using?

For scheduling a meeting could you please write to me at nkarani (at) vision.ee.ethz.ch?

**Comment by mathilde-b** [comment]
Hello, ok I will write to you there ! I solved this but still encountered some issues.
Specifically, the 3D processing of the data is only available for data_hcp_3d. It would useful to have data_prostate_3d.py as well, and other examples also. Thanks !

**Comment by baraaHassan** [comment]
Hello Mathilde, were you able to reproduce it with Tensorflow2 @mathilde-b ?
